### PR TITLE
Fix virtual machine hangs after 2x suspend vm breakpoint

### PR DESCRIPTION
**Pull request by raghucssit**
We need to initiate equal number of resume for suspend vm requests by threads.

Fixes https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/320

### CODE REVIEW

**Comment by raghucssit** [comment]
Changed author as per my eclipse account.
Rebased with master.

File: org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIDebugTarget.java [hunk 1]
```diff
@@ -334,6 +334,10 @@ public class JDIDebugTarget extends JDIDebugElement implements
 	 */
 	private final Map<Long, String> objectLabels = new HashMap<>();
 
+	private int suspendCount;
```
Comment (iloveeclipse) [inline org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIDebugTarget.java:337]: Debugger is always multi-threaded, so all variables should be volatile

File: org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIDebugTarget.java [hunk 2]
```diff
@@ -1340,6 +1344,13 @@ protected void resume(boolean fireNotification) throws DebugException {
 			resumeThreads();
 			VirtualMachine vm = getVM();
 			if (vm != null) {
+				if (pendingResumeRequests) {
+					while (suspendCount > 1) {
```
Comment (iloveeclipse) [inline org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIDebugTarget.java:1348]: `suspendCount`, once set from zero to `> 1`, will stay always `>= 1` after this while loop.
Why not check `while (suspendCount > 0)`?

File: org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIDebugTarget.java [hunk 3]
```diff
@@ -1340,6 +1344,13 @@ protected void resume(boolean fireNotification) throws DebugException {
...
+						vm.resume();
+						suspendCount--;
+						pendingResumeRequests = false;
+					}
+				}
```
Comment (iloveeclipse) [inline org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIDebugTarget.java:1353]: Shouldn't `vm.resume();` call below be in the `else` block?

**Review by iloveeclipse — CHANGES_REQUESTED** [review]
Hard to play with / verify the patch without test case.

Please add a test case to `org.eclipse.jdt.debug.tests.breakpoints.SuspendVMBreakpointsTests`

File: org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIDebugTarget.java (hunk 3 above)
Comment (raghucssit) [reply org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIDebugTarget.java:1353]: This is a default `vm.resume();` which was existing.
When the `suspendCount  = 1` this works by default.
If there were more than 1 Suspend VM breakpoints are set they will be resumed inside while loop.
Whenever a `IJavaBreakpoint.SUSPEND_VM` breakpoint is hit `JDIDebugTarget#suspendedByBreakpoint(JavaBreakpoint breakpoint, boolean queueEvent, EventSet set)` is invoked.
So I am counting the number of Suspend VM breakpoints hit. And whenever VM(Debug Target) Resume is requested all the pending VM suspend will be cleared inside while loop.

File: org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIDebugTarget.java (hunk 2 above)
Comment (iloveeclipse) [reply org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIDebugTarget.java]: I've fixed that in my last pushed commit.

**Review by iloveeclipse — APPROVED** [review]

**Comment by raghucssit** [comment]
> @raghucssit : I've reworked PR a bit, but the main idea is same & original fix was in right place, so main idea untouched.
> 
> * get rid of the extra boolean flag
> * renamed counter and changed it to a thread safe AtomicInteger
> * added comments in code
> * slightly changed test
> 
> Please review.

Code changes looks good.
I have pulled and did some manual testing, Everything works as expected.
Good to merge.

**Pull request merged**